src/storage-factory.js

```diff
@@ -0,0 +1,51 @@
+//
+// Copyright 2020 DxOS.
+//
+
+import assert from 'assert';
+
+import { STORAGE_NODE, STORAGE_CHROME, STORAGE_MOZILLA, STORAGE_IDB } from './storage-types';
+
+const isNode = typeof process !== 'undefined' && process.versions != null && process.versions.node != null;
+const isBrowser = typeof window !== 'undefined';
+
+const defaultStorageType = () => {
+  if (isNode) {
+    return STORAGE_NODE;
+  }
+
+  if (isBrowser) {
+    if (window.requestFileSystem || window.webkitRequestFileSystem) {
+      return STORAGE_CHROME;
+    }
+
+    if (window.IDBMutableFile) {
+      return STORAGE_MOZILLA;
+    }
+
+    return STORAGE_IDB;
+  }
+
+  throw new Error('Unknown platform');
+};
+
+export const createStorageFactory = storageTypes => (root, type = defaultStorageType()) => {
+  assert(typeof root === 'string');
+  assert(storageTypes[type], `Invalid type: ${type}`);
+
+  const storage = new storageTypes[type](root);
+
+  function randomAccessStorage (file) {
+    return storage.create(file);
+  }
+
+  // TODO(burdon): We should have a wrapper rather than adding properties to other objects.
```

Comment:
I don't understand this comment, that object is created by us and must be a function and the only way to add more methods is this one.

Reply:
It isn't the only way since the factory can return anything we want. It's not important for now, but since this issue comes up a lot let's discuss in our 1:1.

Reply:
> It isn't the only way since the factory can return anything we want.
It depends, if you want to be compatible with random-access-storage then you don't have choice, it must be a function.